### PR TITLE
[Core] Show_webui segfault fix.

### LINKED ISSUE #8309: [Async actor/Dashboard] 0x1020a3790 __pyx_pw_3ray_7_raylet_10CoreWorker_13set_webui_display()
### What is the problem?

*Ray version and other system information (Python version, TensorFlow version, OS):*
Nightly

When I ran ray microbenchmark in the latest master, I could see these errors that breaks the dashboard. I assume it happens when async actor monitor is calling the show_web_ui.
Screen Shot 2020-05-04 at 10 54 44 AM

### Reproduction (REQUIRED)
Please provide a script that can be run to reproduce the issue. The script should have **no external library dependencies** (i.e., use fake or mock data / environments):

Run ray microbenchmark. It happens sometimes at async actor benchmark.

- [ ] I have verified my script runs in a clean environment and reproduces the issue.
- [ ] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/latest/installation.html).

### PR DESCRIPTION
**Pull request by rkooo567**
## Why are these changes needed?

This will resolve https://github.com/ray-project/ray/issues/8309

The problem was that CoreWorker removes its cpp instance before default_worker.py's shutdown is called (where threads are joined). That means that async thread's coroutines are still running and using CoreWorker that was already GC'ed. This PR will make sure those coroutines are cleaned properly before CoreWorker is GC'ed.

This rarely happens in sync cases as well (https://github.com/ray-project/ray/issues/8291) due to import threads (look at Simon's comment). It is so much more rare than async cases because there is a mechanism to "stop thread" before they are joined on shutdown. I didn't handle this in this PR because of two reasons. 

- It hasn't been issues for a long time (although I am sure it was there from before). This is very rare (it is detected only once out of 8 long running tests for 3 days whereas async cases occur almost every microbenchmark script run) and fixing this won't probably have big impact compared to efforts required.
- Fixing this will require uncertain amount of refactoring because we need to change the cleanup mechanism. Moving thread join logic to CoreWorker didn't work. It will break Java code as well because we need to run Python code inside Cpp code. So, I decided to deprioritize this for now. I will keep the other issue open (https://github.com/ray-project/ray/issues/8291) so I can revisit later.
This has been manually tested. I ran the reproducible script 10 times, and verified the error didn't occur. It didn't properly handle this issue yet https://github.com/ray-project/ray/issues/8291 because it was a bit tricky to fix. 

## Related issue number

https://github.com/ray-project/ray/issues/8309

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: python/ray/_raylet.pyx [hunk 1]
```diff
@@ -634,6 +634,12 @@ cdef shared_ptr[CBuffer] string_to_buffer(c_string& c_str):
                 <uint8_t*>(c_str.data()), c_str.size(), True))
 
 
+cdef void terminate_asyncio_thread() nogil:
+    with gil:
+        core_worker = ray.worker.global_worker.core_worker
+        core_worker.destory_event_loop_if_exists()
```
Comment (edoakes) [inline python/ray/_raylet.pyx:640]:
```suggestion
        core_worker.destroy_event_loop_if_exists()
```

**Review by edoakes — COMMENTED** [review]
Nice! Thanks for tracking this down. I do think it makes sense to also fix the other issue you mentioned for the sync case if it won't take too much effort.

**Review by edoakes — APPROVED** [review]

File: python/ray/_raylet.pyx (hunk 1 above)
Comment (rkooo567) [reply python/ray/_raylet.pyx:640]: Fxied!

**Comment by rkooo567** [comment]
@edoakes I will assign the task to me and revisit maybe at the end of the sprint. It doesn't look really straightforward. I might need some time to think about it. 

Also, I will cherry-pick this change to the next release. Do you think I should re-run some of tests again with this? (maybe just serve long running test)?

**Pull request merged**